### PR TITLE
reduce bitswap lock contention

### PR DESCRIPTION
**Pull request by Stebalien**
Guard the ledgerMap with a rwlock. This way, we can parallelize notifying bitswap of new blocks.

### CODE REVIEW

File: exchange/bitswap/decision/engine.go [hunk 1]
```diff
@@ -270,8 +270,8 @@ func (e *Engine) addBlock(block blocks.Block) {
 }
 
 func (e *Engine) AddBlock(block blocks.Block) {
-	e.lock.Lock()
-	defer e.lock.Unlock()
+	e.lock.RLock()
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply exchange/bitswap/decision/engine.go]: I believe so. It only reads the `ledgerMap` and the `peerRequestQueue` is threadsafe.

[the remark this answers is not in the document]
Comment (Kubuxu) [reply exchange/bitswap/decision/engine.go]: Cool

**Comment by whyrusleeping** [comment]
I think we should hold off on this until after the release. The potential risk is high, despite us being pretty sure this is a reasonable thing to do.

**Comment by Kubuxu** [comment]
@Stebalien what do you think of using `sync.Map` here? Reading its description it doesn't seem a great fit but maybe?
https://golang.org/pkg/sync/#Map

**Comment by Stebalien** [comment]
> what do you think of using sync.Map here? Reading its description it doesn't seem a great fit but maybe?

IIRC, we end up with some race conditions due to some issues with how we handle connect/disconnect notifications. Once we get the libp2p refactor through, I think we'll be able to do this.

**Comment by schomatis** [comment]
The function `findOrCreate` could be used in `PeerConnected` and `PeerDisconnected` to take advantage of the new `RWMutex`.

Regarding `addBlock`, could a function be introduced that returns all the current ledgers from `ledgerMap` (with a read lock), to iterate over that and release the `ledgerMap` lock earlier in the function? The function `peerRequestQueue.Push` called inside the `for` seems to be expensive to wait on it. The risk is that unncesary work may be done on an already disconnected `peer.ID` (that was removed from the `ledgerMap`).

Also, and this is just a question (not strictly related to this PR), the difference between the exported `AddBlock` and the internal `addBlock` is the lock taking, but is there a situation where `addBlock` could be called alone (without lock)? Having the functionality distributed in this way makes it less evident that `addBlock` needs to use the `ledgerMap` lock.

**Comment by Stebalien** [comment]
Bitswap has been extracted and largely reworked.

**Pull request closed without merging**